docs/puzzles/day17.md

```diff
@@ -2,10 +2,410 @@ import Solver from "../../../../website/src/components/Solver.js"
+  `step` to the probe, and replace `maxY` by the maximum of `maxY` and
+  the `y` position of the current probe,
+- we will ignore any iteration steps where the `probe` does not collide
+  with, or go beyond the target, i.e. the probe is still on a valid
+  trajectory,
+- we then find the first iteration that is not ignored, meaning that at
+  that step the probe must have either collided with the target, or gone
+  beyond it.
+- we then extract the `maxY` of that iteration, provided that the `probe`
+  collided with the target.
+
+The function `simulate` implements the rules above, taking parameters
+`probe` and `target`, and returning an optional value -
+`Some(maxY)` if the trajectory reaches the target, and `None` if it went
+beyond:
+
+```scala
+def simulate(probe: Probe, target: Target): Option[Int] =
+  LazyList
+    .iterate((probe, 0))((probe, maxY) => (step(probe), maxY `max` probe.position.y))
+    .dropWhile((probe, _) => !collides(probe, target) && !beyond(probe, target))
+    .headOption
+    .collect { case (probe, maxY) if collides(probe, target) => maxY }
+```
+> The above code uses `LazyList.iterate`: it creates an infinite sequence
+> of steps, applied in sequence to an initial value, where each step is
+> evaluated on-demand. Next we call `dropWhile`, acting like a condition
+> of a while loop - it limits the size of our sequence because we know
+> that eventually one of the conditions will be broken.
+> We then call `headOption` to get an `Option` wrapping the first element
+> we are interested in. Finally `collect` allows us to inspect `probe`
+> and `maxY`, and keep `maxY` when `probe` matches the condition we want,
+> otherwise if the condition is not met `None` will be returned.
+
+#### Checking all Possible Trajectories
+
+So far we have seen how to simulate the trajectory of a single
+probe. We need to find the best possible height reached by all
+trajectories - meaning that we will need to generate some initial
+velocities for the probe.
+
+We can use some knowledge to help us reduce the search space for
+possible velocities.
+
+First, we assume that the target will always be in a positive direction
+from the probe's initial direction, and that the probe's `x` velocity will
+only get closer to `0`, so we will not need to consider
+negative `x` velocities.
+
+Second, we know that the problem requires us to find the highest positive
+height reached by the probe, and that the probe's `y` velocity can only
+fall once in motion. So we will not need to consider negative `y`
+velocities.
+
+That gives us the lower bounds for `x` and `y` velocities, what about
+the upper bound for the velocities?
+
+We know that a trajectory is invalid if it goes beyond the target in a
+single step, so the largest single step that the probe can move is the
+distance of the furthest corner of the target. For our problem that is the
+lower right corner, so our upper bounds will be the position of that corner.
```

Comment:
This explanation lacks the critical detail that, for *positive* initial y velocities, eventually there will be a point where the position is on the x axis (y position is 0) with a y velocity of `-initialYVelocity - 1`. This is why we can use `-bottomRightY` as upper bound for the y velocity.

Reply:
I have now addressed this comment